oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java

```diff
@@ -144,6 +161,64 @@ public static Builder implicit() {
 		return new Builder(AuthorizationGrantType.IMPLICIT);
 	}
 
+	/**
+	 * Returns a new {@link Builder}, initialized with the values
+	 * from the provided {@code authorizationRequest}.
+	 *
+	 * @since 5.1
+	 * @param authorizationRequest the authorization request used for initializing the {@link Builder}
+	 * @return the {@link Builder}
+	 */
+	public static Builder from(OAuth2AuthorizationRequest authorizationRequest) {
+		Assert.notNull(authorizationRequest, "authorizationRequest cannot be null");
+
+		return new Builder(authorizationRequest.getGrantType())
+				.authorizationUri(authorizationRequest.getAuthorizationUri())
+				.clientId(authorizationRequest.getClientId())
+				.redirectUri(authorizationRequest.getRedirectUri())
+				.scopes(authorizationRequest.getScopes())
+				.state(authorizationRequest.getState())
+				.additionalParameters(authorizationRequest.getAdditionalParameters());
+	}
+
+	/**
+	 * Returns a {@code URI} representation for the provided {@code authorizationRequest}.
+	 *
+	 * @since 5.1
+	 * @param authorizationRequest the authorization request
+	 * @return a {@code URI} representation of the authorization request
+	 */
+	public static URI asUri(OAuth2AuthorizationRequest authorizationRequest) {
```

Comment:
This method does not make sense to me. What is it constructing? The OAuth2AuthorizationRequest has a authorizationRequestUri property, so it is illogical to pass in an argument that contains the value that you are producing. I think this logic belongs encapsulated within the resolver and we can open it up later if necessary.

Reply:
> The OAuth2AuthorizationRequest has a authorizationRequestUri property, so it is illogical to pass in an argument that contains the value that you are producing

Yeah good point

> I think this logic belongs encapsulated within the resolver

Why not keep this logic in `OAuth2AuthorizationRequest` encapsulated. It provides the default `authorizationRequestUri` in `build()` if the user has not supplied it.

```
			authorizationRequest.authorizationRequestUri =
					this.authorizationRequestUri != null ?
						this.authorizationRequestUri : asUri(authorizationRequest);

```